starboard/shared/starboard/player/filter/audio_discard_duration_tracker.h

```diff
@@ -0,0 +1,103 @@
+// time of the playback when the audio buffers do not support partial audio.
+class AudioDiscardDurationTracker {
+ public:
+  // Stores the timestamps and audio discard durations of an InputBuffer
+  // in an AudioDiscardInfo. |buffer_length| is given to determine the timestamp
+  // of when audio begins discarding from the back of an InputBuffer.
+  // Each audio discard duration creates a separate AudioDiscardInfo, so there
+  // may be two AudioDiscardInfos for one InputBuffer if audio is discarded from
+  // both the front and back.
+  // When audio is discarded from the front,
+  // discard_start_timestamp == input_buffer->timestamp(), while
+  // discard_end_timstamp == input_buffer->timestamp()
+  // + <audio discarded from the front>
+  // When audio is discarded from the back,
+  // discard_start_timestamp == input_buffer->timestamp() + |buffer_length|
+  // - <audio discarded from the back>, while discard_end_timstamp
+  // == input_buffer->timestamp() + |buffer_length|.
+  // When the sum of audio discarded from the front and back matches or exceeds
+  // the length of the buffer, a single AudioDiscardInfo is created to discard
+  // the entire length of the buffer.
+  void CacheDiscardDuration(const scoped_refptr<InputBuffer>& input_buffer,
+                            SbTime buffer_length);
+  void CacheMultipleDiscardDurations(const InputBuffers& input_buffers,
+                                     SbTime buffer_length);
+  // Adjusts the given timestamp by taking the total length of discarded audio
+  // into account. When |timestamp| falls within the current AudioDiscardInfo,
+  // AdjustTimeForTotalDiscardDuration() will return a timestamp based on the
+  // start timestamp of the discard duration. When |timestamp| is beyond the end
+  // of the discard duration, the discard duration will be added to the running
+  // total. This way AdjustTimeForTotalDiscardDuration() will "freeze" the
+  // timestamp until the end of the discard duration.
+  //
+  // Ex. Given a |timestamp| 28 ms with |total_discard_duration| == 0 and
+  // current AudioDiscardInfo == { discard_duration: 10 ms,
+  // discard_start_timestamp: 22 ms, discard_end_timestamp: 32 ms},
+  // AdjustTimeForTotalDiscardDuration() returns discard_start_timestamp -
+  // |total_discard_duration| == 22 ms. If a |timestamp| 33 ms is passed later,
+  // discard_duration is added to |total_discard_duration|, and the
+  // AudioDiscardInfo is removed.
+  //
+  // Timestamps passed to AdjustTimeForTotalDiscardDuration() must be
+  // monotonically increasing for the algorithm to run correctly.
+  SbTime AdjustTimeForTotalDiscardDuration(SbTime timestamp);
+  void Reset() {
+    discard_infos_ = std::queue<AudioDiscardInfo>();
+    total_discard_duration_ = 0;
+    last_received_timestamp_ = 0;
+  }
+
+ private:
+  struct AudioDiscardInfo {
+    SbTime discard_duration;
+    SbTime discard_start_timestamp;
+    SbTime discard_end_timestamp;
+  };
+
```

Comment:
Consider adding ThreadChecker here (a lesson learnt from b/289281412).

Reply:
I added a mutex since `AdjustTimeForTotalDiscardDuration()` is called from `GetCurrentMediaTime()`, so it's expected to be called on a different thread than the other methods.